rfcs/1-input-pipeline.md

```diff
@@ -0,0 +1,396 @@
+for example to assert that `X` and `y` are of the same length or that `y` is purely numeric for regressors.
+SciKeras also validates and inspects the model, for example to make sure that the output shape matches the
+target's shape.
+These checks are helpful for simple models, but may be too restrictive for more complex scenarios, like
+multi-input/output models.
+
+This RFC proposes a unifified interface for composable data transformations and validations.
+The goal is to provide a pipeline of default transformations and validations that cover the simple use cases,
+while allowing users to easily remove checks or add transformation steps for more advanced use cases.
+
+Some of the speicific functional requirements are:
+1. Able to implement the default (i.e. current) data transformations and validations. This includes:
+ - Integer encoding targets for classifiers.
+ - One-hot encoding targets for classifiers using the categorical crossentropy loss.
+ - Converting class probability predictions into class predictions for classifiers.
+2. Able to implement user-defined transformations, including:
+ - Splitting the input and/or target into multiple inputs/outputs.
+ - Reshaping 2D inputs into 3D.
+3. Able to operate on array-like data (lists, Numpy arrays, Pandas DataFrames, etc.) as well as `tf.data.Dataset`s.
+4. Composable and modifiable without subclassing.
+
+## Proposal
+
+This proposal consists of an expansion / generalization of our current `BaseWrapper.target_encoder` and `BaseWrapper.feature_encoder` data pipeline concepts to encompass:
+1. Manipulation of the entire dataset (X, y & sample_weight) both as array-like data and as `tf.data.Dataset`.
+2. A standard interface to establish a reference to the current `BaseWrapper` instance (i.e. a setter for `model`).
+
+Additionally, this proposal seeks to establish a contract on where in the dataflow these pipelines are executed.
+
+These pipelines will consist of chained transformers implementing a Scikit-Learn-like interface, but without being restricted to the exact Scikit-Learn API.
+
+For example:
+
+```python
+from typing import Any, Dict, Optional, Tuple, Sequence, Union, TYPE_CHECKING
+
+import numpy as np
+import tensorflow as tf
+
+from scikeras.wrappers import BaseWrapper
+
+
+Data = Tuple[np.ndarray, Union[np.ndarray, None], Union[np.ndarray, None]]
+ArrayLike = Union[Sequence, np.ndarray]
+
+
+class NotInitializedError(Exception):
+    ...
+
+
+class ArrayTransformer:
+
+    def set_model(self, model: "BaseWrapper") -> None:
+        self.model = model
+
+    def transform_input(self, X: ArrayLike, y: Optional[ArrayLike], sample_weight: Optional[ArrayLike], *, initialize: bool = True) -> Data:
+        return X, y, sample_weight
+
+    def transform_output(self, y_pred_proba: np.ndarray, y: Union[np.ndarray, None]) -> Tuple[np.ndarray, Union[np.ndarray, None]]:
+        return y_pred_proba, None
+
+    def get_meta(self) -> Dict[str, Any]:
```

Comment:
:-1:

Why not have an attribute `meta_`?

Reply:
The `get_meta` is stuck in there from older revision.

My current thoughts are to not have any interface for these meta parameters and instead let the implementers set attributes on `model` directly (or I guess we could have a `meta_` attribute on `BaseWrapper` and then do some `__getattr__` stuff to expose these attributes).

Sorry for having old stuff left in there.